File: spikes/classifier/python/judge.py
"""Spike — `judge` (ADR 0020 D4), Python port idiom. THROWAWAY.

``judge(on=..., ask=..., rule=bands(...))`` -> a Judge whose ``as_guardrail()``
returns the shape ``python/src/toolnexus/agents/loop.py:guarded_hooks`` already
consumes: a ``(event) -> str`` callable where ``""``/``"allow"`` means allow and any
other string is the deny reason.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, Literal, Mapping, Optional

from classifier import Classifier, Question, ScoreAnswer

Band = Literal["allow", "ask", "deny"]


@dataclass
class Verdict:
    kept: Band
    evidence: dict[str, Any] = field(default_factory=dict)
    model: str = ""
    calibrated: bool = True
    reason: str = ""


# --------------------------------------------------------------------------- rules

Rule = Callable[["Judge", Any, Mapping[str, Any]], Verdict]


def bands(key: str, cuts: list[tuple[float, Band]], *, escalate: Optional[dict[str, float]] = None) -> Rule:
    """`bands` rule: put one score answer in a band; ``escalate`` lifts allow->ask when
    a noul answer exceeds its threshold. First matching cut wins (cuts are ordered)."""

    def rule(judge: "Judge", state: Any, questions: Mapping[str, Question]) -> Verdict:
        decision = judge.classifier.evaluate(state, questions)
        ans = decision.answers[key]
        assert isinstance(ans, ScoreAnswer)
        kept: Band = cuts[-1][1]
        for upper, band in cuts:
            if ans.score < upper:
                kept = band
                break
        evidence: dict[str, Any] = {key: ans.score, f"{key}.confidence": ans.confidence}
        for nkey, threshold in (escalate or {}).items():
            got = getattr(decision.answers[nkey], "noul", 0.0)
            evidence[nkey] = got
            if got >= threshold and kept == "allow":
                kept = "ask"
        reason = "" if kept == "allow" else f"{kept}: {key}={ans.score} ({_level(ans, ans.score)})"
        return Verdict(kept=kept, evidence=evidence, model=decision.model, reason=reason)

    return rule


def _level(ans: ScoreAnswer, score: float) -> str:
    return ans.legend.get(str(int(round(score))), "")


# --------------------------------------------------------------------------- judge


@dataclass
class Judge:
    classifier: Classifier
    on: Callable[[Any], Any]
    ask: Mapping[str, Question]
    rule: Rule
    fail_closed: bool = True

    def verdict(self, event: Any) -> Verdict:
        try:
            return self.rule(self, self.on(event), self.ask)
        except Exception as exc:  # fail-closed: an unreachable classifier denies
            if not self.fail_closed:
                return Verdict(kept="allow")
            return Verdict(kept="deny", reason=f"deny: classifier unavailable ({type(exc).__name__})")

    def as_guardrail(self, only: tuple[Band, ...] = ("ask", "deny")) -> Callable[[Any], str]:
        """-> the port's Guardrail shape. A judge can only move a call toward ask/deny:
        it never returns anything that could widen an earlier denial."""

        def guardrail(event: Any) -> str:
            v = self.verdict(event)
            return v.reason if v.kept in only else ""

        return guardrail


def judge(*, on: Callable[[Any], Any], ask: Mapping[str, Question], rule: Rule, fail_closed: bool = True):
    def build(classifier: Classifier) -> Judge:
        return Judge(classifier=classifier, on=on, ask=ask, rule=rule, fail_closed=fail_closed)

    return build

File: spikes/classifier/elixir/lib/judge.ex
defmodule Spike.Judge do
  @moduledoc """
  `judge/1` turns a Classifier into a guardrail: the same `(event -> verdict)`
  shape the shipped port already composes with first-deny-wins
  (`Toolnexus.Agents.Loop.guarded_hooks/2`, `elixir/lib/toolnexus/agents/loop.ex:147`).

  Options:
    * `:on`      — tool name(s) this judge applies to; anything else is allowed untouched
    * `:ask`     — `%{key => Question}`, the questions put to the classifier
    * `:state`   — `(event -> term())`, what the classifier is shown
    * `:rule`    — `%{key => [{upper_bound, verdict}, ...]}` bands, upper-exclusive,
                   evaluated in order; verdict is `:allow | :ask | :deny`
    * `:backend` — `(request -> Decision)`
  """

  alias Spike.Classifier
  alias Spike.Classifier.{NoulAnswer, ScoreAnswer}

  @type verdict :: :allow | :ask | :deny

  @doc "Evaluate and return `{verdict, reason}`."
  @spec decide(keyword(), map()) :: {verdict(), String.t()}
  def decide(opts, event) do
    if applies?(opts[:on], event) do
      decision =
        opts[:backend].(
          Classifier.request(opts[:state].(event), opts[:ask])
        )

      opts[:rule]
      |> Enum.map(fn {key, bands} -> band(bands, decision.answers[key], key) end)
      |> Enum.max_by(&rank(elem(&1, 0)))
    else
      {:allow, ""}
    end
  end

  @doc "The guardrail-shaped form: `\"\"` ⇒ allow, a reason string ⇒ denied."
  @spec guardrail(keyword()) :: (map() -> String.t())
  def guardrail(opts) do
    fn event ->
      case decide(opts, event) do
        {:allow, _} -> ""
        {_, reason} -> reason
      end
    end
  end

  defp applies?(nil, _event), do: true
  defp applies?(on, event) when is_list(on), do: event[:tool] in on
  defp applies?(on, event), do: event[:tool] == on

  defp rank(:allow), do: 0
  defp rank(:ask), do: 1
  defp rank(:deny), do: 2

  # Bands read the numeric face of an answer; the struct says which field that is.
  defp band(bands, %ScoreAnswer{score: v} = a, key), do: pick(bands, v, key, legend(a, v))
  defp band(bands, %NoulAnswer{noul: v}, key), do: pick(bands, v, key, nil)

  defp legend(%ScoreAnswer{legend: l}, v) when is_map(l),
    do: l[Integer.to_string(round(v))]

  defp legend(_, _), do: nil

  defp pick([{upper, verdict} | rest], value, key, label) do
    if upper == :infinity or value < upper do
      {verdict, reason(verdict, key, value, label)}
    else
      pick(rest, value, key, label)
    end
  end

  defp reason(:allow, _key, _value, _label), do: ""

  defp reason(verdict, key, value, label) do
    detail = if label, do: " — #{label}", else: ""
    "#{verdict}: #{key} #{fmt(value)}#{detail}"
  end

  defp fmt(v) when is_float(v), do: Float.to_string(v)
  defp fmt(v), do: to_string(v)
end
